src/modules/0.0.22/aws_s3/mappers/bucket_object.ts

```diff
@@ -0,0 +1,130 @@
+import { S3, _Object, paginateListObjectsV2, waitUntilObjectNotExists } from '@aws-sdk/client-s3';
+import { WaiterOptions } from '@aws-sdk/util-waiter';
+
+import { AwsS3Module } from '..';
+import { AWS, crudBuilder2, crudBuilderFormat, paginateBuilder } from '../../../../services/aws_macros';
+import { Context, Crud2, MapperBase } from '../../../interfaces';
+import { Bucket, BucketObject } from '../entity';
+
+export class BucketObjectMapper extends MapperBase<BucketObject> {
+  module: AwsS3Module;
+  entity = BucketObject;
+
+  async bucketObjectMapper(instance: _Object, ctx: Context, bucket: String, region: string) {
+    const bo: BucketObject = new BucketObject();
+    if (!instance.Key) throw new Error('Received a bucket object without a key');
+
+    // check for bucket instance
+    if (!Object.values(ctx.memo?.cloud?.Bucket ?? {}).length) {
+      bo.bucket =
+        (await this.module.bucket.db.read(ctx, `${bucket}|${region}`)) ??
+        (await this.module.bucket.cloud.read(ctx, `${bucket}|${region}`));
+    } else {
+      bo.bucket =
+        (await this.module.bucket.db.read(ctx, `${bucket}|${region}`)) ??
+        ctx?.memo?.cloud?.BucketAWS?.[`${bucket}|${region}`];
+    }
+
+    bo.region = region;
+    bo.eTag = instance.ETag;
+    bo.key = instance.Key;
+    return bo;
+  }
+
+  equals = (a: BucketObject, b: BucketObject) => {
+    const res = Object.is(a.eTag, b.eTag);
+    return res;
+  };
+
+  getBucketObjects = paginateBuilder<S3>(paginateListObjectsV2, 'Contents', undefined, undefined, (b, k) => ({
+    Bucket: b,
+    Prefix: k,
+  }));
+
+  deleteBucketObject = crudBuilder2<S3, 'deleteObject'>('deleteObject', (b, k) => ({ Bucket: b, Key: k }));
+
+  cloud = new Crud2<BucketObject>({
+    create: async (es: BucketObject[], ctx: Context) => {
+      // we cannot create buckets
+      const out: any = [];
+      return out;
+    },
+    read: async (ctx: Context, id?: string) => {
+      const enabledRegions = (await ctx.getEnabledAwsRegions()) as string[];
+
+      if (!!id) {
+        const { bucketName, objectKey, region } = this.idFields(id);
```

Comment:
I do not think this might be working. For the `bucketName` is what I said about the `@cloudId` for the multiple join column, not sure what it is the property that it is being decorated. The, in the entity the property decorated is `key`, not `objectKey`

Reply:
We might need to think about how to stress these cases in the test suite for coverage and to avoid future errors when we relate this entity with another one.